File: test_auto_add.py
"""
Test automatizado para verificar funcionalidad auto-add
"""
import time
import requests
import json

def test_aplicacion():
    print("🧪 INICIANDO TESTS AUTOMATIZADOS\n")

    # Test 1: Verificar que la aplicación esté corriendo
    print("📡 TEST 1: Conectividad con aplicación")
    try:
        response = requests.get('http://localhost:5000', timeout=5)
        if response.status_code == 200:
            print("✅ Aplicación accesible en localhost:5000")
        else:
            print(f"❌ Aplicación responde con código: {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Error conectando con aplicación: {e}")
        return False

    # Test 2: Verificar endpoint de sugerencias
    print("\n🔍 TEST 2: API de sugerencias")
    try:
        response = requests.get('http://localhost:5000/sugerencias', timeout=5)
        if response.status_code == 200:
            data = response.json()
            print("✅ API sugerencias responde correctamente")
            print(f"📊 Datos recibidos: {len(data.get('medidas', []))} medidas, {len(data.get('marcas', []))} marcas")

            # Mostrar sample de datos
            if data.get('medidas'):
                print(f"🔸 Medidas ejemplo: {', '.join(data['medidas'][:3])}...")
            if data.get('marcas'):
                print(f"🔸 Marcas ejemplo: {', '.join(data['marcas'][:3])}...")

            return data
        else:
            print(f"❌ API sugerencias falla con código: {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Error en API sugerencias: {e}")
        return False

def test_validacion():
    print("\n📝 TEST 3: Lógica de validación")

    # Simular datos de entrada
    test_data = {
        "medida": "175/65 R14",
        "cantidad": "2",
        "marca": "MICHELIN",
        "neto": "45.50",
        "ganancia": "",  # Vacío para test
        "ecotasa": "",   # Vacío para test
        "iva": ""        # Vacío para test
    }

    print("🧪 Datos de prueba:")
    for key, value in test_data.items():
        status = "✅" if value else "❌"
        print(f"   {key}: \"{value}\" {status}")

    # Simular validación básica (4 campos)
    validacion_basica = all([
        test_data["medida"],
        test_data["cantidad"],
        test_data["marca"],
        test_data["neto"]
    ])

    # Simular validación completa (7 campos)
    validacion_completa = all([
        test_data["medida"],
        test_data["cantidad"],
        test_data["marca"],
        test_data["neto"],
        test_data["ganancia"],
        test_data["ecotasa"],
        test_data["iva"]
    ])

    print(f"\n📋 VALIDACIÓN BÁSICA (4 campos): {'✅ PASS' if validacion_basica else '❌ FAIL'}")
    print(f"📋 VALIDACIÓN COMPLETA (7 campos): {'✅ PASS' if validacion_completa else '❌ FAIL'}")

    if validacion_basica and not validacion_completa:
        print("⚠️ ESTADO ESPERADO: Validación básica pasa, completa falla")
        print("💡 RESULTADO: Auto-add debería funcionar con solo 4 campos")
        return True
    else:
        print("❌ PROBLEMA: Validación no está como se esperaba")
        return False

def main():
    print("=" * 60)
    print("🧪 TEST AUTOMATIZADO AUTO-ADD FUNCTIONALITY")
    print("=" * 60)

    # Ejecutar tests
    sugerencias_data = test_aplicacion()
    if not sugerencias_data:
        print("\n❌ Tests de conectividad fallaron")
        return

    validacion_ok = test_validacion()
    if not validacion_ok:
        print("\n❌ Tests de validación fallaron")
        return

    print("\n" + "=" * 60)
    print("✅ TODOS LOS TESTS AUTOMATIZADOS PASARON")
    print("=" * 60)

    print("\n🎯 PRÓXIMOS PASOS MANUALES:")
    print("1. Abre http://localhost:5000 en tu navegador")
    print("2. Presiona F12 > Consola")
    print("3. Ejecuta: testAutoAdd()")
    print("4. Prueba manualmente:")
    print("   - Rellena: medida, cantidad=2, neto=30")
    print("   - Click en chip de marca")
    print("   - Verificar que se añade automáticamente")

    print(f"\n📊 DATOS DISPONIBLES PARA PRUEBAS:")
    if isinstance(sugerencias_data, dict):
        if sugerencias_data.get('medidas'):
            print(f"   Medidas: {len(sugerencias_data['medidas'])} disponibles")
        if sugerencias_data.get('marcas'):
            print(f"   Marcas: {len(sugerencias_data['marcas'])} disponibles")

if __name__ == "__main__":
    main()

File: test_auto_add.js
// Test de funcionalidad de auto-añadir sugerencias
// Ejecutar en la consola del navegador en http://localhost:5000

console.log("🧪 Iniciando test de auto-añadir sugerencias...");

function testAutoAddFunctionality() {
  console.log("\n📋 Test: Verificando elementos necesarios");

  // Verificar que existen los elementos necesarios
  const elements = {
    medida: document.getElementById("presupuesto-medida"),
    cantidad: document.getElementById("presupuesto-cantidad"),
    ganancia: document.getElementById("presupuesto-ganancia"),
    ecotasa: document.getElementById("presupuesto-ecotasa"),
    iva: document.getElementById("presupuesto-iva"),
    marcaTemp: document.getElementById("presupuesto-marca-temp"),
    netoTemp: document.getElementById("presupuesto-neto-temp"),
    btnAgregarMarca: document.getElementById("btnAgregarMarca"),
    sugMarcas: document.querySelector("#sugerencias-marcas"),
    sugMedidas: document.querySelector("#sugerencias-medidas"),
  };

  let missingElements = [];
  for (const [name, element] of Object.entries(elements)) {
    if (!element) {
      missingElements.push(name);
      console.log(`❌ ${name}: No encontrado`);
    } else {
      console.log(`✅ ${name}: Encontrado`);
    }
  }

  if (missingElements.length > 0) {
    console.log(`\n❌ Faltan elementos: ${missingElements.join(", ")}`);
    return false;
  }

  console.log("\n✅ Todos los elementos necesarios están presentes");
  return true;
}

function fillTestData() {
  console.log("\n📝 Rellenando campos de prueba...");

  // Rellenar todos los campos excepto marca
  const testData = {
    "presupuesto-medida": "205/55/16 91v",
    "presupuesto-cantidad": "2",
    "presupuesto-ganancia": "25",
    "presupuesto-ecotasa": "3",
    "presupuesto-iva": "21",
    "presupuesto-neto-temp": "30",
  };

  for (const [id, value] of Object.entries(testData)) {
    const element = document.getElementById(id);
    if (element) {
      element.value = value;
      element.dispatchEvent(new Event("input"));
      console.log(`✅ ${id}: ${value}`);
    }
  }

  // Actualizar Materialize si está disponible
  if (window.M) {
    M.updateTextFields();
  }

  console.log("✅ Campos de prueba rellenados");
}

function testChipClick() {
  console.log("\n🖱️ Test: Simulando click en chip de marca...");

  // Buscar chips de marcas disponibles
  const marcaChips = document.querySelectorAll("#sugerencias-marcas .chip");

  if (marcaChips.length === 0) {
    console.log("❌ No se encontraron chips de marcas. Primero carga las sugerencias.");
    return false;
  }

  console.log(`✅ Encontrados ${marcaChips.length} chips de marcas`);

  // Simular click en el primer chip
  const firstChip = marcaChips[0];
  console.log(`🖱️ Haciendo click en: "${firstChip.textContent}"`);

  firstChip.click();

  // Verificar que se rellenó el campo marca
  setTimeout(() => {
    const marcaInput = document.getElementById("presupuesto-marca-temp");
    if (marcaInput && marcaInput.value) {
      console.log(`✅ Campo marca rellenado con: "${marcaInput.value}"`);

      // Verificar si se añadió automáticamente
      setTimeout(() => {
        // Aquí podrías verificar si se añadió a la lista de neumáticos
        console.log("🔍 Verificando si se añadió automáticamente...");
        // Esta verificación dependería de cómo muestres los neumáticos añadidos
      }, 200);
    } else {
      console.log("❌ El campo marca no se rellenó correctamente");
    }
  }, 200);

  return true;
}

function runCompleteTest() {
  console.log("🚀 EJECUTANDO TEST COMPLETO DE AUTO-AÑADIR\n");

  if (!testAutoAddFunctionality()) {
    console.log("❌ Test fallido: Elementos faltantes");
    return;
  }

  fillTestData();

  setTimeout(() => {
    testChipClick();
  }, 500);

  console.log('\n💡 Si no ves chips de marcas, haz click en "Cargar Sugerencias" primero');
}

// Función auxiliar para cargar sugerencias si no están cargadas
function loadSuggestions() {
  const btnCargarSugerencias =
    document.querySelector('[onclick*="cargarSugerencias"]') ||
    document.querySelector('button:contains("Cargar")') ||
    document.getElementById("btnCargarSugerencias");

  if (btnCargarSugerencias) {
    console.log("🔄 Cargando sugerencias...");
    btnCargarSugerencias.click();
    setTimeout(runCompleteTest, 2000);
  } else {
    console.log(
      "⚠️ No se encontró botón para cargar sugerencias. Ejecutando test sin sugerencias..."
    );
    runCompleteTest();
  }
}

// Exportar funciones para uso manual
window.autoAddTest = {
  runCompleteTest,
  testAutoAddFunctionality,
  fillTestData,
  testChipClick,
  loadSuggestions,
};

console.log("\n💡 Funciones disponibles:");
console.log("   autoAddTest.runCompleteTest() - Ejecutar test completo");
console.log("   autoAddTest.loadSuggestions() - Cargar sugerencias y test");
console.log("   autoAddTest.fillTestData() - Solo rellenar campos");
console.log("   autoAddTest.testChipClick() - Solo test de click");

// Auto-ejecutar si se detectan sugerencias cargadas
if (document.querySelectorAll("#sugerencias-marcas .chip").length > 0) {
  setTimeout(runCompleteTest, 1000);
} else {
  setTimeout(loadSuggestions, 1000);
}
